Refactor DPI scaling and fix the GTK implementation.

xStrom opened the pull request:
## Background
This work got started due to the bug report in #862. @makoConstruct ran into an issue with GTK where there would occasionally be a single line of unused space in the window. I confirmed that this was an issue related to DPI scaling. @makoConstruct had a DPI of 75.0 but the issue appears with a bunch of other DPIs too.

## Reproducing the bug
On my Ubuntu 19.10 I could easily reproduce these issues by changing my DPI via [gnome-tweaks](https://askubuntu.com/a/1034071/586162). Changing the text scaling factor is what changes the DPI. Entering `0.78` will roughly match a DPI of 75.0. Other fractional numbers work too. In my testing the GTK DPI handling was completely broken.

## The calculator with `master` and with this PR
![before](https://user-images.githubusercontent.com/754881/81237046-3f561b80-9007-11ea-991a-57165d6c3875.png) ![after](https://user-images.githubusercontent.com/754881/81237062-467d2980-9007-11ea-9682-7debcb4a10bd.png)

## Getting more value out of time
As I already had researched the DPI code thoroughly to understand how to fix GTK, I figured I might as well improve the `druid-shell` DPI scaling system in general.

I added the `Scale` struct, which is a platform independent resolution scaling helper. It can do proper scaling per axis and has a bunch of helper methods for converting between logical and platform pixels.

Previously `druid-shell` was leaking some platform pixels via `WinHandler::size`, which were handled by e.g. `druid`. I moved that into `druid-shell` so that all DPI related work is a `druid-shell` implementation detail. If someone really wants then they can still get information about the platform pixels and the scaling by fetching the `Scale` via `WindowHandle::get_scale`.

I also updated the Windows and web backends to make use of the new `Scale` system.

## Pixel terminology
Druid has been using the web terminology where a pixel and a pixel aren't the same thing. This is confusing to say the least. There is the CSS pixel unit (abbreviated **px**) and the device pixel (abbreviated **px**). I know we've been generally following web naming, but I'm convinced this is not a case where it makes sense. The situation in the web world with px and px isn't the result of some wise design, it's just an overloading of a term in the name of backwards compatibility because all the existing docs/code already use pixel/px. We have a chance to be much less ambiguous and confusing.

Unfortunately there isn't really industry consensus about what to call logical pixels.
- Microsoft's [device independent pixel](https://docs.microsoft.com/en-us/windows/win32/learnwin32/dpi-and-device-independent-pixels) - **dip**
- Google's [density independent pixel](https://material.io/design/layout/pixel-density.html#density-independence) - ~**dip**~ - they're in the process of rebranding as **dp** (pronounced, I kid you not, **dip**)
- Apple's [point](https://developer.apple.com/design/human-interface-guidelines/ios/icons-and-images/image-size-and-resolution/) - **pt**

Are there any others with good abbreviation?

I think it's important that we move away from calling both platform pixels and logical pixels - pixels. I think we should keep platform pixels as pixels and use something else for logical pixels, something that isn't confusing and has a good abbreviation.

The problem with **dip**/**dp** is that for a newcomer it's easy to imagine it meaning *device pixel*. It's not even that wrong, that's what the **d** truly stands for! We want the opposite effect.

The problem with **pt** is that it would be overloading a different existing unit, from the text/print world. It also has a bit of a clash with `Point`.

It's not super clear to me what the best choice is. Suggestions and discussion is welcome on this. What I do know is that I want to stop the use of pixel for both concepts.

For this PR I went with point/**pt**, but it can be changed if we think another choice is superior.

## Future work
This PR addresses the DPI issue on the shell side, but there's still a remaining issue on the druid side. While the shell issue revealed itself as an empty pixel line at the edge, the druid issue reveals itself as a clipped pixel line at the same edges. It has to do with `BoxConstraints` expanding to integers. The solution however isn't as trivial as just removing the expansion, because some other widgets like `Flex` depend on that behavior. A more thorough solution is needed, which is out of scope for this PR here.

Fixes #862

cmyr commented on druid-shell/src/scale.rs:
```diff
@@ -0,0 +1,289 @@
+    /// Converts from points into pixels,
+    /// using the x axis scale factor for `x` and the y axis scale factor for `y`.
+    #[inline]
+    pub fn pt_to_px_xy<T: Into<f64>>(&self, x: T, y: T) -> (f64, f64) {
```
this is a very annoying suggestion from me, _but_: if we had a trait `Scaleable`, which we implemented for our geometry types, we could just have,

```rust
trait Scaleable {
    fn to_display_point(&self, scale: &Scale) -> Self;
    fn to_pixel(&self, scale: &Scale) -> Self;
}

impl Scale {
    fn to_display_point<T: Scaleable>(&self, geom: &T) -> T {
        geom.to_display_point(self)
    }
    // etc
}
```

This would at least reduce the overall API surface?

xStrom commented:
I think having a `Scalable` trait is an improvement and will allow for easier extensions, so I implemented it.

I think *display point* / **dp** is a fine choice, so I changed everything to that. Using `dp` in method names will also allow for some backwards compatible rebranding if the need arises.

cmyr commented on druid-shell/src/mouse.rs:
```diff
@@ -21,9 +21,7 @@ use crate::keyboard::KeyModifiers;
-    /// The location of the mouse in the current window.
-    ///
-    /// This is in px units not device pixels, that is, adjusted for hi-dpi.
+    /// The location of the mouse in display points in relation to the current window.
```
I would include a section in the docs of `scale.rs` that explicitly explains our use of 'pixel/px' and 'display point/dp', and then we can link to that section everywhere we use one of those terms?

It would be nice if we had some canonical doc explanation of the dp/px distinction, and then we could link to that everywhere we use one of those terms? I'd make this super obvious, something with a section in the `//! top level` doc comment called like '## Pixels (px) and display points (dp)'. Or something? I'd just like it to be easy for someone reading docs (more important in druid than druid-shell I guess) to be able to easily access an explanation.

xStrom replied:
Sure, I'll write some more docs about it. I also already wrote a book chapter about it in #913.

cmyr commented on druid-shell/src/platform/web/window.rs:
```diff
@@ -230,16 +221,19 @@ fn setup_resize_callback(ws: &Rc<WindowState>) {
-            .handler
-            .borrow_mut()
-            .size(physical_width, physical_height);
+        let size_dp = state.scale.try_borrow_mut().map_or(None, |mut scale| {
```
if this is converting an error to an option I would use `.ok().map(_)`.

cmyr commented on druid-shell/src/platform/web/window.rs:
```diff
@@ -230,16 +221,19 @@ fn setup_resize_callback(ws: &Rc<WindowState>) {
+            state.canvas.set_height(size_px.height as u32);
+            let _ = state.context.scale(scale.scale_x(), scale.scale_y());
+            Some(scale.size_dp())
+        });
```
does it make sense to just move this to a method on the state type? `try_update_scale` or something? Not sure, just seems like maybe.

xStrom replied:
Sure, why not. I moved it to its own method.

cmyr commented on druid-shell/src/platform/web/window.rs:
```diff
@@ -544,34 +541,16 @@ impl WindowHandle {
+            }
+        } else {
+            Err(ShellError::WindowDropped)
+        }
```
just for fun, I probably write this as:

```rust
self.0
    .upgrade()
    .ok_or(ShellError::WindowDropped)?
    .scale
    .try_borrow()
    .map_err(|_| BorrowError::new("WindowHandle::get_scale", "scale", false).into())
    .map(|s| s.clone())
```

cmyr commented on druid-shell/src/platform/windows/util.rs:
```diff
@@ -103,6 +105,28 @@ impl FromWide for [u16] {
 
+/// Converts a `Rect` to a winapi `RECT`.
+#[inline]
+pub fn rect_to_recti(rect: Rect) -> RECT {
```
`pub(crate)` here, probably; the main advantage is we'll get compiler warnings in the future if these are unused. (Probably not an issue in practice since this module shouldn't be public, but good general strategy I think)

xStrom commented:
Non-uniform x/y scaling is definitely not common. It is supported by drawing, but the platform scale factor provided by Windows/macOS/web is always uniform from what I've gathered. However Linux supports it. So I guess this is mostly a question if we want to support it in Linux. The cost of supporting it isn't that high as it's mostly an implementation detail of `Scale`, especially if you do conversions on thigs that implement `Scalable`.

I haven't really looked into DPI scaling images yet, because I felt this PR was getting big as it is already. However that's definitely the next significant step on the road to DPI success. In general yes the app could provide a set of images and then we should select the smallest that is equal or bigger than the target area.

Having `Scale` in `Env` might be useful. I'll look more into that and see what can be done and if it makes sense.

xStrom commented:
Okay I made changes based on the feedback. I added some more documentation around display points and a bunch of doc links pointing to it. I surfaced the `Scale` type in druid, which also surfaces that documentation.

I looked into adding `Scale` to `Env` but there was no clear way to do it, because `Env` seems global. However `Scale` can be different per window. Not sure if `Env` should grow the capability to have per-window state or if we should have some sort of `Event::Scale(Scale)` event.

I did add the `scale` method to `WinHandler` so the `druid-shell` part of informing changes to this is done. The druid side will take more thinking and currently it will ignore that change. I think it may make more sense to figure out the bitmap scaling story first - as in does scaling images even work right now and how to do it. Then we can figure out how to best surface scale changes.

In that sense I think this PR can just be the shell part, which is done. We can deal with the druid side in a different PR as we have more answers.

cmyr commented on druid-shell/src/scale.rs:
```diff
@@ -0,0 +1,388 @@
+
+    /// Returns `true` if the specified DPI is approximately equal to the `Scale` DPI.
+    #[inline]
+    pub fn dpi_approx_eq(&self, dpi_x: f64, dpi_y: f64) -> bool {
```
is there a specific rationale for this, instead of say making the caller create a `Scale` and having an `approx_eq` for that?

xStrom replied:
The rationale has vanished. This used to be useful with an initial implementation that used a different strategy and attempted to always achieve integer logical display points. That no longer being the case combined with `Scale` now just being the scale factors, we can remove this. The new `Scale` implements `PartialEq` which will be just fine for our needs.

cmyr commented on druid-shell/src/scale.rs:
```diff
@@ -0,0 +1,388 @@
+    /// Converts from pixels into display points,
+    /// using the x axis scale factor for `x` and the y axis scale factor for `y`.
+    #[inline]
+    pub fn px_to_dp_xy<T: Into<f64>>(&self, x: T, y: T) -> (f64, f64) {
```
can't we just rely on the `Scaleable` impl on `Vec2`/`Point` etc?

cmyr commented on druid-shell/src/scale.rs:
```diff
@@ -0,0 +1,388 @@
+
+    /// Converts from pixels into display points, using the x axis scale factor.
+    #[inline]
+    pub fn px_to_dp_x<T: Into<f64>>(&self, x: T) -> f64 {
```
do we anticipate these two methods be used? If it's an obscure case, I would be happy to require the user to construct a point, convert that, and then take the `x`.

cmyr commented on druid-shell/src/scale.rs:
```diff
@@ -0,0 +1,388 @@
+
+    /// Returns the x axis platform DPI associated with this `Scale`.
+    #[inline]
+    pub fn dpi_x(&self) -> f64 {
```
I would probably just have a single `dpi` method that returns either a `(f64, f64)` or `Vec2`.

cmyr commented on druid-shell/src/scale.rs:
```diff
@@ -0,0 +1,388 @@
+    }
+}
+
+impl Scalable for Line {
```
Is this used? I would have thought that scaling is happening to points, sizes, and rects, but not other shapes.

cmyr commented on druid-shell/src/scale.rs:
```diff
@@ -0,0 +1,388 @@
+    }
+}
+
+impl Scalable for Insets {
```
ditto this; do we use insets to store like window chrome size or something?

xStrom commented:
I pushed a new version which fixes the typo and removes some code.
- `dpi_approx_eq` removed because it's not really needed anymore.
- `px_to_dp_xy` / `px_to_dp_x` / `px_to_dp_y` were adaptations of the old DPI methods in `druid-shell`. They're no longer used thanks to `Scalable` and so yeah we can delete them.

I think there's value in keeping the `dpi_x` and `dpi_y` methods. They're used right now like this:
```rust
let props = D2D1_RENDER_TARGET_PROPERTIES {
    _type: D2D1_RENDER_TARGET_TYPE_DEFAULT,
    pixelFormat: D2D1_PIXEL_FORMAT {
        format: DXGI_FORMAT_B8G8R8A8_UNORM,
        alphaMode: D2D1_ALPHA_MODE_IGNORE,
    },
    dpiX: scale.dpi_x() as f32,
    dpiY: scale.dpi_y() as f32,
    usage: D2D1_RENDER_TARGET_USAGE_NONE,
    minLevel: D2D1_FEATURE_LEVEL_DEFAULT,
};
```
It's possible to achieve this even with a unified method of course, but it would be an extra line.

`Line` and `Insets` are not used right now, but could be useful if say the X11 backend implements its own menu. I don't think the cost of having them is too high and they provide some completeness.
